url_decoding: Add the `Filter` class.

PieterCK opened the pull request:
Prep PR to fix: #31100. It adds  `class Filter` which is the lighter Python copy of `filter.Filter` in the frontend. The functions that we want out of this class are:
- canonicalization & validation logic to clean up narrow terms
- the `update_terms` method
- the `sorted_terms` method

PR that depends on these changes: #32272
<details>
<summary>Self-review checklist</summary>

<!-- Prior to submitting a PR, follow our step-by-step guide to review your own code:
https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code -->

<!-- Once you create the PR, check off all the steps below that you have completed.
If any of these steps are not relevant or you have not completed, leave them unchecked.-->

- [x] [Self-reviewed](https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code) the changes for clarity and maintainability
      (variable names, code reuse, readability, etc.).

Communicate decisions, questions, and potential concerns.

- [x] Explains differences from previous plans (e.g., issue description).
- [x] Highlights technical choices and bugs encountered.
- [x] Calls out remaining decisions and concerns.
- [x] Automated tests verify logic where appropriate.

Individual commits are ready for review (see [commit discipline](https://zulip.readthedocs.io/en/latest/contributing/commit-discipline.html)).

- [x] Each commit is a coherent idea.
- [x] Commit message(s) explain reasoning and motivation for changes.

Completed manual review and testing of the following:

- [] Visual appearance of the changes.
- [] Responsiveness and internationalization.
- [ ] Strings and tooltips.
- [ ] End-to-end functionality of buttons, interactions and flows.
- [ ] Corner cases, error conditions, and easily imagined bugs.
</details>

PieterCK committed:
url_decoding: Add the base `Filter` class.

This is the Python copy of a class of the same name in frontend's
`filter.Filter` class. Although it roughly follows the same set
validations, it probably won't have the same set of methods with the
frontend's `Filter`. It does not validate the referenced objects(users
and channels), and validates terms differently--e.g, converts operand to
prefered types when possible--, adjusting for its use in the backend.

The server version will primarily be used to fix narrow links in
exported messages during import.

PieterCK committed:
url_decoding: Add `Filter.operands` method.

The Python copy of a class method with the same name in frontend's
`filter.Filter`.

PieterCK committed:
url_decoding: Add `Filter.fix_redundant_is_private` method.

The Python copy of a class method with the same name in frontend's
`filter.Filter`.

timabbott commented on zerver/lib/url_decoding.py:
```diff
@@ -152,3 +154,48 @@ def parse_narrow_url(
+        return terms
+
+    def setup_filter(self, terms: Sequence[NarrowTerm]) -> list[NarrowTerm]:
+        return self.fix_terms(terms)
```
This seems fine in concept, but it'd be good to extend this to something that has tests.

PieterCK replied:
Yeah, I think the methods that we really want from this class are `generate_redirect_url` and a new method to update a specific term, right? something like `update_term` that accepts a `Mapping[NarrowTerm, NarrowTerm]` where the key is supposed to be the current term you want to replace and the value is the updated version. 

I have a series of commits building up to those methods, I wonder if there are any other methods we want to include in this.

PieterCK replied:
I ended up making a set of "generate narrow URL" functions instead of creating a copy of `generate_redirect_url`. Turns out that function doesn't do what I initially thought--turn the list of terms into a narrow URL.

PieterCK committed:
narrow_helper: Create a `NarrowTermOperandT` type alias.

This is a prep commit to add the `Filter.operands` method which could
use the same type alias.

PieterCK commented on zerver/lib/url_decoding.py:
```diff
@@ -187,6 +189,50 @@ def term_type(term: NarrowTerm) -> str:
+                return 999
+
+        def compare(a: str, b: str) -> int:
+            return level(a) - level(b)
```
In the frontend, the comparison function for sorting included a logic to do local-aware sorting in the case where there are no differences between the two item:

```javascript
        const compare = (a: string, b: string): number => {
            const diff = level(a) - level(b);
            if (diff !== 0) {
                return diff;
            }
            return util.strcmp(a, b);
        };

        return [...term_types].sort(compare);
```

I'm not sure why we need to do local-aware sort here, especially in the backend version. Though if we for whatever reason need to implement that in the server, the [Python implementation](https://docs.python.org/3/howto/sorting.html#comparison-functions) of it looks like it is simpler than the frontend

PieterCK committed:
url_decoding: Add `Filter.update_term` method.

This adds a new method to update a specific term in a Filter instace.
This could be used in the operation to fix narrow links in
exported messages during import.

PieterCK commented on zerver/lib/url_decoding.py:
```diff
@@ -337,3 +342,70 @@ def update_term(self, existing_term: NarrowTerm, new_term: NarrowTerm) -> bool:
         return True
+
+    def generate_narrow_url(self) -> str:
+        """
```
I've not gone through the trouble of making this able to guess and build the correct narrow URL out of the given narrow terms. This only builds what I assume are the most common kinds of narrow URLs one would find in exported messages, but we can extend it when the need arises.

Or would it be better if there are separate explicit functions building specific types of narrow URLs? Either way during import we probably know what type of narrow link we're passing into `prase_narrow_url` and  `Filter`

PieterCK commented on zerver/lib/url_decoding.py:
```diff
@@ -141,3 +149,101 @@ def parse_narrow_url(
+        self._setup_filter(terms)
+
+    @staticmethod
+    def canonicalize_term(term: NarrowTerm) -> NarrowTerm:
```
I know we don't need to look up the referenced object IDs here, but I'm not sure how strict or thorough this should be, for now, it has a lot more validations and cleaning logic than frontend's `canonicalize_term`--it includes some more validations from `narrow.NarrowBuilder`

PieterCK replied:
Also let me know if it's better to use `if else` instead of `match case` here :eyes:. I used it because it can do simple type-checking in a concise manner(less calling `issinstace()`).

PieterCK committed:
url_decoding: Add `Filter.get_terms` method.

This adds a utility method to get specific NarrowTerms from a Filter
instance. This could be used in the operation to fix narrow links in
exported messages during import.

PieterCK commented on zerver/lib/url_decoding.py:
```diff
@@ -141,3 +147,120 @@ def parse_narrow_url(
+                pass
+
+            case _:
+                raise BadNarrowOperatorError(f"unknown '{operator}' operand {operand!s}")
```
In the frontend we don't do this validation, instead:
```javascript
    switch (operator) {
            ...
            case:
            ...
            default:
                operand = operand.toString().toLowerCase();
        }
```

PieterCK commented on zerver/lib/url_decoding.py:
```diff
@@ -141,3 +154,322 @@ def parse_narrow_url(
+                return operand
+            case str():
+                if operand == "":
+                    raise BadNarrowOperatorError("Invalid 'channel' operand")  # no coverage
```
This fixture tests this logic:
```json
    "channel_operator_validations": {
        ...
        "invalid_terms": [
            { "operator": "channel", "operand": [1, 2, 3], "negated": false },
            { "operator": "channel", "operand": "", "negated": false }
        ]
    },
```

PieterCK commented on zerver/lib/url_decoding.py:
```diff
@@ -141,3 +154,322 @@ def parse_narrow_url(
+                if operand.isdigit():
+                    return [int(operand)]
+                if operand == "":
+                    raise BadNarrowOperatorError("Invalid user ID")  # no coverage
```
THis fixture tests this logic:
```json
"dm_operator_validations": {
        ...
        "invalid_terms": [
            { "operator": "dm", "operand": ["1", "2", "3"], "negated": false },
            { "operator": "dm", "operand": [], "negated": false },
            { "operator": "dm", "operand": "", "negated": false },
            { "operator": "dm", "operand": ["totoro@zulip.com", "nausicaa@zulip.com", "ponyo@zulip.com"], "negated": false }
        ]
    },
```

PieterCK commented on zerver/lib/url_decoding.py:
```diff
@@ -141,3 +154,322 @@ def parse_narrow_url(
+            case "dm-including" | "sender" if isinstance(operand, str | int):
+                operand = int(operand) if str(operand).isdigit() else str(operand).lower()
+                if operand == "":
+                    raise BadNarrowOperatorError("Invalid user ID")  # no coverage
```
This fixture tests this logic:
```json
"dm_including_operator_validations": {
       ...
        "invalid_terms": [
            { "operator": "dm-including", "operand": [1, 2, 3], "negated": false },
            { "operator": "dm-including", "operand": "", "negated": false }
        ]
    },
```

PieterCK commented on zerver/lib/url_decoding.py:
```diff
@@ -141,3 +154,322 @@ def parse_narrow_url(
+                return operand
+            case str():
+                if operand.strip() == "":
+                    raise BadNarrowOperatorError("Invalid 'channel' operand")  # nocoverage
```
This fixture tests this logic:
```json
    "channel_operator_validations": {
        ...
        "invalid_terms": [
            { "operator": "channel", "operand": [1, 2, 3], "negated": false },
            { "operator": "channel", "operand": " ", "negated": false }
        ]
    },
```

PieterCK replied:
This fixture tests this logic:
```json
    "channel_operator_validations": {
        ...
        "invalid_terms": [
            { "operator": "channel", "operand": [1, 2, 3], "negated": false },
            { "operator": "channel", "operand": " ", "negated": false }
        ]
    },
```

PieterCK commented on zerver/lib/url_decoding.py:
```diff
@@ -141,3 +154,322 @@ def parse_narrow_url(
+                if operand.isdigit():
+                    return [int(operand)]
+                if operand.strip() == "":
+                    raise BadNarrowOperatorError("Invalid user ID")  # nocoverage
```
THis fixture tests this logic:
```json
"dm_operator_validations": {
        ...
        "invalid_terms": [
            { "operator": "dm", "operand": ["1", "2", "3"], "negated": false },
            { "operator": "dm", "operand": [], "negated": false },
            { "operator": "dm", "operand": " ", "negated": false },
            { "operator": "dm", "operand": ["totoro@zulip.com", "nausicaa@zulip.com", "ponyo@zulip.com"], "negated": false }
        ]
    },
```

PieterCK commented on zerver/lib/url_decoding.py:
```diff
@@ -141,3 +154,322 @@ def parse_narrow_url(
+        # In practice this shouldn't happen, no-op here means its either an empty
+        # string("") or empty list([]) operand which we've filtered out in
+        # `reformat_dm_operand`.
+        raise BadNarrowOperatorError(  # nocoverage
```
This fixture tests this logic:
```json
    "channel_operator_validations": {
        ...
        "invalid_terms": [
            { "operator": "channel", "operand": [1, 2, 3], "negated": false },
            { "operator": "channel", "operand": " ", "negated": false }
        ]
    },
```

PieterCK commented:
Marking this as ready for review

PieterCK commented:
I tried updating #32272 to use the `Filter` class, and it feels like it has all the right methods and validations for that operation. That is to say, #32272 is likely no longer blocked. But I'll continue to work on #32272 once this has received at least a round of review first.

PieterCK commented:
Update: Previously `generate_dm_with_url` calls `full_name = user_profiles[0].get_username()` to get a users fullname, which is wrong. I updated it to call `full_name = user_profiles[0].full_name`.

```diff
diff --git a/zerver/lib/url_decoding.py b/zerver/lib/url_decoding.py
index 96c4862c3c..cb01f61e3e 100644
--- a/zerver/lib/url_decoding.py
+++ b/zerver/lib/url_decoding.py
@@ -442,7 +442,7 @@ class Filter:
                     user_ids=users_ids_or_emails,
                     realm=self._realm,
                 )
-                full_name = user_profiles[0].get_username()
+                full_name = user_profiles[0].full_name
                 pm_slug = str(users_ids_or_emails[0]) + "-" + hash_util_encode(full_name)
                 return f"/#narrow/dm/{pm_slug}"
```

PieterCK commented:
Updated the `generate_*_url` functions to generate fragments-only narrow links, as per discussed here [CZO](https://chat.zulip.org/#narrow/channel/378-api-design/topic/We're.20generating.20relative.20links.20that.20doesn't.20start.20with.20.22.2F.22.2E/with/2130985)

PieterCK committed:
url_decoding: Add `Filter.sorted_terms` method.

This sorts narrow terms using the same algorithm
`filter.Filter.sorted_term_types` uses.

This is a prep commit to dynamically construct a narrow link out of a
list of narrow terms. Ideally, the terms in narrow links is sorted from
general to specific, so we have this method to take care of that.

PieterCK commented:
@zulipbot add "mentor review"

hmm, would it make sense for you to review this PR? @chdinesh1089 
This is one of the prep PR for #32272. I've tried to write a summary of the project and where this PR sits in the scheme of things here : [#backend > Fix narrow URLs in exported messages.](https://chat.zulip.org/#narrow/channel/3-backend/topic/Fix.20narrow.20URLs.20in.20exported.20messages.2E/with/2208096). Let me know if any part is unclear!

PieterCK commented on zerver/lib/url_decoding.py:
```diff
@@ -141,3 +148,211 @@ def parse_narrow_url(
+
+        return sorted(terms, key=cmp_to_key(compare))
+
+    def _build_sorted_term_types(self) -> list[str]:
```
Ah, this is left over from a previous version of this PR, currently we don't have any use for this, so I'm dropping the `_build_sorted_term_types` method.

PieterCK commented:
@chdinesh1089 Thanks for taking a look at this, I've addressed your reviews! I think this is ready for you to take a look again.

chdinesh1089 commented on zerver/tests/test_url_decoding.py:
```diff
@@ -79,3 +79,32 @@ def test_get_operands(self) -> None:
+            NarrowTerm(negated=False, operator="topic", operand="testing"),
+            NarrowTerm(negated=False, operator="topic", operand="testing2"),
+            NarrowTerm(negated=False, operator="near", operand="1"),
+            NarrowTerm(negated=False, operator="near", operand="2"),
```
I think it's better to have tests be comprehensive but this is probably OK.

PieterCK commented:
@zulipbot add "maintainer review"

PieterCK commented:
Failing CI is:
 `2025-09-25 15:25:57 [documentation_crawler] ERROR: Please check link: https://zulip.com/help/protect-your-account`
